src/nias_cpp/interfaces/vectorarray.h

```diff
@@ -353,6 +369,71 @@ class VectorArrayInterface
         axpy(std::vector<F>{alpha}, x, indices, x_indices);
     }
 
+    /**
+     * \brief Returns a const reference to the i-th vector
+     *
+     * \note Not all vector array implementations give access to the underlying vectors,
+     * e.g., the NumpyVectorArray does not. In this case, the method will throw a NotImplementedError.
+     */
+    [[nodiscard]] virtual const VectorInterface<F>& vector(ssize_t /*i*/) const
+    {
+        throw nias::NotImplementedError("No access to underlying vectors.");
+    }
+
+    /**
+     * \brief Returns a mutable reference to the i-th vector
+     *
+     * \note Not all vector array implementations give access to the underlying vectors,
+     * e.g., the NumpyVectorArray does not. In this case, the method will throw a NotImplementedError.
+     */
+    [[nodiscard]] virtual VectorInterface<F>& vector(ssize_t /*i*/)
+    {
+        throw nias::NotImplementedError("No access to underlying vectors.");
+    }
+
+    /**
+     * \brief Returns a const reference to the vector at index \c i cast to the specified type \c VectorType.
+     *
+     * \tparam VectorType The type to which the vector should be cast. Must derive from VectorInterface<F>.
+     * \param i The index of the vector to retrieve.
+     * \throws InvalidArgumentError if the vector cannot be cast to \c VectorType.
+     * \throws NotImplementedError if the vector array implementation does not provide random vector access
+     * \sa vector(ssize_t)
+     */
+    template <class VectorType>
+        requires std::derived_from<VectorType, VectorInterface<F>>
+    [[nodiscard]] const VectorType& vector_as(ssize_t i) const
+    {
+        try
+        {
+            return dynamic_cast<const VectorType&>(vector(i));
+        }
+        catch (std::bad_cast&)
+        {
+            throw InvalidArgumentError("vector at index " + std::to_string(i) + " is not of type " +
+                                       typeid(VectorType).name());
+        }
+    }
```

Comment:
As discussed , I think this is better off in the vector interface itself. This would require to write
```c++
vector_array.vector(i)._as<MyType>()
```
but I think that's ok. I've put in https://github.com/nias-project/nias-cpp/issues/24 to change this later. Just resolve this after reading @tobiasleibner